packages/cardano-services/src/Asset/TypeormAssetProvider/TypeormAssetProvider.ts

```diff
@@ -0,0 +1,64 @@
+import { Asset, AssetProvider, Cardano, GetAssetArgs, ProviderError, ProviderFailure } from '@cardano-sdk/core';
+import { AssetEntity } from '@cardano-sdk/projection-typeorm';
+import { NftMetadataService } from '../types';
+import { TypeormProvider, TypeormProviderDependencies } from '../../util';
+
+interface TypeormAssetProviderDependencies extends TypeormProviderDependencies {
+  nftMetadataService: NftMetadataService;
+}
+
+export class TypeormAssetProvider extends TypeormProvider implements AssetProvider {
+  #dependencies: TypeormAssetProviderDependencies;
+
+  constructor(dependencies: TypeormAssetProviderDependencies) {
+    const { connectionConfig$, entities, logger } = dependencies;
+    super('TypeormAssetProvider', { connectionConfig$, entities, logger });
+
+    this.#dependencies = dependencies;
+  }
+
+  async getAsset({ assetId, extraData }: GetAssetArgs): Promise<Asset.AssetInfo> {
+    const assetInfo = await this.#getAssetInfo(assetId);
+
+    if (extraData?.nftMetadata) assetInfo.nftMetadata = await this.#getNftMetadata(assetInfo);
+
+    return assetInfo;
+  }
+
+  async getAssets(): Promise<Asset.AssetInfo[]> {
+    // todo: implement getting multiple assets
+    return [];
+  }
+
+  async #getNftMetadata(asset: Asset.AssetInfo): Promise<Asset.NftMetadata | null> {
+    return this.#dependencies.nftMetadataService.getNftMetadata({
+      name: asset.name,
+      policyId: asset.policyId
+    });
```

Comment:
We still want to return base AssetInfo when we fail to get nft metadata
```suggestion
  async #getNftMetadata(asset: Asset.AssetInfo): Promise<Asset.NftMetadata | null | undefined> {
    try {
      return this.#dependencies.nftMetadataService.getNftMetadata({
        name: asset.name,
        policyId: asset.policyId
      });
    } catch (error) {
      this.logger.error('Failed to get nft metadata', asset.assetId, error);
      return;
    }
```